sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvcompLZ4CompressionCodec.scala

```diff
@@ -0,0 +1,155 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import ai.rapids.cudf.{BaseDeviceMemoryBuffer, ContiguousTable, Cuda, DeviceMemoryBuffer}
+import ai.rapids.cudf.nvcomp.{BatchedLZ4Compressor, BatchedLZ4Decompressor, CompressionType, Decompressor, LZ4Compressor}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.format.{BufferMeta, CodecType}
+
+/** A table compression codec that uses nvcomp's LZ4-GPU codec */
+class NvcompLZ4CompressionCodec extends TableCompressionCodec with Arm {
+  override val name: String = "nvcomp-LZ4"
+  override val codecId: Byte = CodecType.NVCOMP_LZ4
+
+  override def compress(
+      tableId: Int,
+      contigTable: ContiguousTable): CompressedTable = {
+    val tableBuffer = contigTable.getBuffer
+    val (compressedSize, oversizedBuffer) = NvcompLZ4CompressionCodec.compress(tableBuffer)
+    closeOnExcept(oversizedBuffer) { oversizedBuffer =>
+      val tableMeta = MetaUtils.buildTableMeta(
+        tableId,
+        contigTable.getTable,
+        tableBuffer,
+        CodecType.NVCOMP_LZ4,
+        compressedSize)
+      CompressedTable(compressedSize, tableMeta, oversizedBuffer)
+    }
+  }
+
+  override def decompressBuffer(
+      outputBuffer: DeviceMemoryBuffer,
+      outputOffset: Long,
+      outputLength: Long,
+      inputBuffer: DeviceMemoryBuffer,
+      inputOffset: Long,
+      inputLength: Long): Unit = {
+    withResource(outputBuffer.slice(outputOffset, outputLength)) { outSlice =>
+      withResource(inputBuffer.slice(inputOffset, inputLength)) { inSlice =>
+        NvcompLZ4CompressionCodec.decompress(outSlice, inSlice)
+      }
+    }
+  }
+
+  override def createBatchCompressor(maxBatchMemoryBytes: Long): BatchedTableCompressor = {
+    new BatchedNvcompLZ4Compressor(maxBatchMemoryBytes)
+  }
+
+  override def createBatchDecompressor(maxBatchMemoryBytes: Long): BatchedBufferDecompressor = {
+    new BatchedNvcompLZ4Decompressor(maxBatchMemoryBytes)
+  }
+}
+
+object NvcompLZ4CompressionCodec extends Arm {
+  // TODO: Make this a config?
+  val LZ4_CHUNK_SIZE: Int = 64 * 1024
+
+  /**
+   * Compress a data buffer.
+   * @param input buffer containing data to compress
+   * @return the size of the compressed data in bytes and the (probably oversized) output buffer
+   */
+  def compress(input: DeviceMemoryBuffer): (Long, DeviceMemoryBuffer) = {
+    val tempSize = LZ4Compressor.getTempSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE)
+    withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+      var compressedSize: Long = 0L
+      val outputSize = LZ4Compressor.getOutputSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+        tempBuffer)
+      closeOnExcept(DeviceMemoryBuffer.allocate(outputSize)) { outputBuffer =>
```

Comment:
any change our size is off where we would want to check afterwards to make sure it fit in the buffer (in either temp or output?

Reply:
There is no way I can verify the temporary storage usage since it's usage is completely opaque to the caller.

I can add a sanity check on the output size.

Reply:
I added sanity checks on the output size.